File: backend/app/locales.py
"""Języki interfejsu — jedno miejsce, z którego korzystają backend i front.

Polski jest językiem BAZOWYM: w nim powstają teksty i on jest zapasem, gdy dla innego
języka brakuje tłumaczenia. Interfejs pokazuje kody ISO 639-1 (`PL`, `EN`).

Skąd bierze się język zalogowanej osoby, w tej kolejności:

1. `users.locale` — wybór konkretnego konta (NULL = brak wyboru),
2. `DEFAULT_LOCALE` — domyślny dla wdrożenia (ZCO chodzi po polsku, demo anglojęzyczne
   mogłoby startować po angielsku bez zmiany kodu),
3. polski.

Języka PRZEGLĄDARKI świadomie nie pytamy. Na komputerze na oddziale konto bywa wspólne
albo przeglądarka ustawiona przez kogoś innego, więc interfejs zmieniałby język między
zmianami bez niczyjej decyzji.
"""

import os

# Kod bazowy. Nie jest „jednym z" — dla pozostałych języków jest wartością zapasową.
BASE_LOCALE = "pl"

# Kolejność ma znaczenie: tak samo idą przyciski w przełączniku.
SUPPORTED_LOCALES: tuple[str, ...] = ("pl", "en")


def normalize_locale(value: str | None) -> str | None:
    """Kod w postaci używanej w bazie albo None, gdy nie rozpoznajemy.

    Przyjmujemy zapisy, które przychodzą z przeglądarek i nagłówków — `EN`, `en-US`,
    `en_GB` — i sprowadzamy je do samego języka. Wartości spoza listy odrzucamy,
    zamiast zapisywać: kolumna `locale` steruje wyborem katalogu tłumaczeń i wpis
    typu `de` bez katalogu zostawiłby użytkownika z pustym interfejsem.
    """
    if not value:
        return None
    kod = value.strip().replace("_", "-").split("-")[0].lower()
    return kod if kod in SUPPORTED_LOCALES else None


def default_locale() -> str:
    """Domyślny język wdrożenia. Błędna wartość w środowisku nie może wywrócić startu."""
    return normalize_locale(os.getenv("DEFAULT_LOCALE")) or BASE_LOCALE

File: frontend/src/i18n/locales.ts
/**
 * Języki interfejsu — lista i reguły wspólne dla całego frontu.
 *
 * Polski jest BAZOWY: w nim powstają teksty i on jest zapasem, gdy w innym języku
 * brakuje tłumaczenia (scalanie katalogów robi `request.ts`). Dzięki temu ekran,
 * do którego jeszcze nie doszliśmy, pokazuje polskie zdanie, a nie klucz w rodzaju
 * `files.uploadButton` — użytkownik widzi wtedy coś zrozumiałego, a tłumacz od razu
 * wie, czego brakuje.
 *
 * Odpowiednik po stronie serwera: `backend/app/locales.py`.
 */

/** Kod bazowy. Nie jest „jednym z" — dla pozostałych jest wartością zapasową. */
export const BASE_LOCALE = 'pl';

/** Kolejność ta sama, co przycisków w przełączniku. Kody ISO 639-1.
 *  Bazowy pierwszy, potem angielski (nim pokazujemy system), dalej alfabetycznie. */
export const LOCALES = ['pl', 'en', 'cs', 'de', 'es', 'uk'] as const;

export type Locale = (typeof LOCALES)[number];

/** Podpis w przełączniku i w `<html lang>`. Nazwa własna języka, nie tłumaczona. */
export const LOCALE_NAMES: Record<Locale, string> = {
  pl: 'Polski',
  en: 'English',
  cs: 'Čeština',
  de: 'Deutsch',
  es: 'Español',
  uk: 'Українська',
};

/** Nazwa ciasteczka, z którego układ główny czyta język przy renderowaniu na serwerze. */
export const LOCALE_COOKIE = 'locale';

/**
 * Języki WŁĄCZONE na tym wdrożeniu (`UI_LANGUAGES`, np. `pl,en`).
 *
 * Po co osobna lista obok `LOCALES`: katalog tłumaczeń może istnieć, zanim będzie
 * kompletny. Wdrożenie klienckie ma wtedy zostać jednojęzyczne — przy jednej pozycji
 * przełącznik znika sam, bo nie ma między czym wybierać.
 *
 * Język bazowy jest zawsze w środku: to on jest zapasem dla wszystkich pozostałych.
 * WYŁĄCZNIE po stronie serwera — `process.env` w przeglądarce tych wartości nie ma.
 */
export function enabledLocales(): Locale[] {
  const zSrodowiska = (process.env.UI_LANGUAGES || '')
    .split(',')
    .map((k) => normalizeLocale(k))
    .filter((k): k is Locale => k !== null);
  const lista = zSrodowiska.length ? zSrodowiska : [...LOCALES];
  // Kolejność zawsze z `LOCALES`, żeby przyciski nie skakały przy zmianie zmiennej.
  return LOCALES.filter((k) => k === BASE_LOCALE || lista.includes(k));
}

export function isLocale(value: unknown): value is Locale {
  return typeof value === 'string' && (LOCALES as readonly string[]).includes(value);
}

/**
 * Kod sprowadzony do postaci z listy albo `null`.
 *
 * Przyjmujemy zapisy przychodzące z przeglądarek i nagłówków — `EN`, `en-US`,
 * `en_GB` — bo ciasteczko bywa ustawiane też ręcznie. Wartości spoza listy
 * odrzucamy: nie ma dla nich katalogu tłumaczeń.
 */
export function normalizeLocale(value: string | null | undefined): Locale | null {
  if (!value) return null;
  const kod = value.trim().replace('_', '-').split('-')[0].toLowerCase();
  return isLocale(kod) ? kod : null;
}
